### PR TITLE
Proposal

### PR DESCRIPTION
**Pull request by hassox**
I've put together a proposal to get the ball rolling on this one. Love some feedback on it.

### CODE REVIEW

**Comment by scrogson** [comment]
@hassox at first glance this looks really fantastic! I'll be working through it today/tomorrow at some point. Great work!

File: mix.exs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  #
+  #   {:mydep, "~> 0.3.0"}
+  #
+  # Or git/path repositories:
+  #
+  #   {:mydep, git: "https://github.com/elixir-lang/mydep.git", tag: "0.1.0"}
+  #
+  # Type "mix help deps" for more examples and options
+  defp deps do
+    [
```
Comment (scrogson) [inline mix.exs]: Let's tidy this up:

```elixir
defp deps do
  [{:plug, "~> 1.0"}]
end
```

File: test/test_helper.exs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Code.require_file "./support/mixins.exs", __DIR__
+Code.require_file "./support/spec_router.exs", __DIR__
+Code.require_file "./support/simple_callback.exs", __DIR__
+Code.require_file "./support/redirector.exs", __DIR__
+Code.require_file "./support/default_options.exs", __DIR__
+Code.require_file "./support/with_errors.exs", __DIR__
+
+
+ExUnit.start()
+Application.ensure_all_started(:bypass)
```
Comment (scrogson) [inline test/test_helper.exs]: I'm not seeing `bypass` in the `deps`.

File: README.md [hunk 3]
```diff
@@ -2,6 +2,112 @@
 
 > An Elixir Authentication System for Plug-based Web Applications
 
+Ueberauth is an authentication framework that is heavily inspired by [Omniauth](https://github.com/intridea/omniauth)
+I would call it a port but it is significantly different in operation - but almost the same by concept. Huge hat tip to omniauth.
```
Comment (scrogson) [inline README.md]: Maybe `s/I would/You could/` ?

**Comment by doomspork** [comment]
This is really exciting stuff guys!

@hassox I saw some typos while reading through the docs, do you want them pointed out or is that annoying?

**Comment by hassox** [comment]
@doomspork I'd love feedback mate :) I'm sorry but typos happen to me all the time :(

**Comment by doomspork** [comment]
Me too!  Sometimes it seems my fingers and brain experience some packet loss :stuck_out_tongue:

File: README.md [hunk 4]
```diff
@@ -2,6 +2,112 @@
...
+
+Strategies are plugs that decorate or intercept requests (or both).
+Strategies implement the two phases and then may allow the request to flow through to your downstream plugs.
+Implementing the request and callback phases is optional, and if not implemented the request will be decorated Ueberauth information
+and allowed to carry on through the pipeline.
+
+### Request Phase
+
+The request phase is where you request information about the user. This could be a redirect to an OAuth2 authorization url
+or a form for collecting username and password. The request phase may be very asynchronous and should be concered with
```
Comment (doomspork) [inline README.md]: "concered with" => "concerned with".

Also is "very asynchronous" a thing?

File: README.md [hunk 5]
```diff
@@ -2,6 +2,112 @@
...
+only the collection of information. When a request comes in on the request phase url the relevant strategy will receive the `request_phase!` call.
+
+In some cases (default) the application using Ueberauth is responsible for implementing the request phase.
+That is, you should setup a route to receive the request phase and provide a form etc.
+In some cases, like OAuth, the request phase is used to redirect your user to a 3rd party site to fullfil the request.
```
Comment (doomspork) [inline README.md]: "fulfill the request"

File: README.md [hunk 6]
```diff
@@ -2,6 +2,112 @@
...
+
+For example, an OAuth strategy for Github will receive the request phase url and stop the request, redirecting you to Githubs OAuth challenge url with some query parameters.
```
Comment (doomspork) [inline README.md]: "GitHub's"

File: README.md [hunk 7]
```diff
@@ -2,6 +2,112 @@
...
+Once you complete the Github OAuth flow, the user will be redirected back to the host site to the callback URL.
+
+Another example is simple email/password authentication.
+A request is made by the client to the request phase path and the host application displays a form.
+The strategy will likely not do anything with the incomming request_phase request and simply pass through to the application.
```
Comment (doomspork) [inline README.md]: "incoming"

File: README.md [hunk 8]
```diff
@@ -2,6 +2,112 @@
...
+
+In phoenix setup a pipeline:
+
+    pipeline :ueberauth do
+      Ueberauth.plug "/auth"
+    end
+
+Its url matching is done via pattern matching rather than explicit runtime checks so your strategies will only fire for relevant requests.
+
+Now that you have this, your strategies will intercept relevant requests for each strategy for both request and callback phases. The default urls are (for our facebook and github example)
```
Comment (doomspork) [inline README.md]: "Facebook" & "GitHub"

File: README.md [hunk 9]
```diff
@@ -2,6 +2,112 @@
...
+Example:
+
+    providers: [
+      identity: { Ueberauth.Strategies.Identity, [ request_path: "/absolute/path", callback_path: "/absolute_path" ] }
+    ]
+
+
+#### Http Methods
+
+By default, all callback urls are only avaialble via the GET method. You can override this via options to your strategy.
```
Comment (doomspork) [inline README.md]: "available"

File: lib/ueberauth.ex [hunk 10]
```diff
@@ -0,0 +1,226 @@
+defmodule Ueberauth do
```
Comment (doomspork) [inline lib/ueberauth.ex]: Typos from the README are also present in the moduledocs.

File: lib/ueberauth/auth.ex [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+              credentials: Credentials.t,
+              extra: Extra.t
+             }
+
+  defstruct uid: nil, # An identifier unique to the given provider, such as a Twitter user ID. Should be stored as a string.
+            provider: nil, # The provider name as defined in the configuration.
+            strategy: nil, # The strategy module used
+            info: %Info{}, # An info struct about the user. See Ueberauth.Auth.Info
+            credentials: %Credentials{}, # A struct containing information on the credentials. See Ueberauth.Auth.Credentials
+            extra: %Extra{} # Any additional information that may be avaialble. See Ueberauth.Auth.Extra
```
Comment (doomspork) [inline lib/ueberauth/auth.ex]: "available"

File: lib/ueberauth/failure/error.ex [hunk 12]
```diff
@@ -0,0 +1,13 @@
+defmodule Ueberauth.Failure.Error do
+  @moduledoc """
+  A specific error for a failed authentication attempt.
+  The message_key may be used to identify fields or other machine interperated methods like translation
```
Comment (doomspork) [inline lib/ueberauth/failure/error.ex]: "interpreted"

File: lib/ueberauth/strategy.ex [hunk 13]
```diff
@@ -0,0 +1,260 @@
+defmodule Ueberauth.Strategy do
+  @moduledoc """
+  The Stragey is the work-horse of the system.
```
Comment (doomspork) [inline lib/ueberauth/strategy.ex]: "Strategy"

File: lib/ueberauth/strategy.ex [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+            Credentials,
+            other: %{
+              password: conn.params["password"],
+              password_confirmation: conn.params["password_confirmation"]
+            }
+          )
+        end
+      end
+
+  After the strategy has runn through the `callback_phase!` function, since there are no errors added, Ueberauth will add the constructed auth struct to the connection.
```
Comment (doomspork) [inline lib/ueberauth/strategy.ex]: "run"

File: lib/ueberauth/strategy.ex [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+          provider: strategy_name(conn),
+          strategy: strategy(conn),
+          uid: uid(conn),
+          info: info(conn),
+          extra: extra(conn),
+          credentials: credentials(conn)
+        )
+      end
+
+  Each component of the struct is a seperate function and receives the connection object. From this Ueberauth will construct and assign the struct for processing in your own controller.
```
Comment (doomspork) [inline lib/ueberauth/strategy.ex]: "separate"

File: lib/ueberauth/strategy.ex [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+  @doc """
+  Provides the extra params for the user. This is one of the component functions that is used to construct the auth struct. What you return here will be in the auth struct at the `extra` key.
+
+  You would include any additional information within extra that does not fit in either `info` or `credentials`
+  """
+  @callback extra(Plug.Conn.t) :: Extra.t
+
+  @doc """
+  Provides the credentals for the user. This is one of the component functions that is used to construct the auth struct. What you return here will be in the auth struct at the `credentials` key.
```
Comment (doomspork) [inline lib/ueberauth/strategy.ex]: "credentials"

**Comment by hassox** [comment]
thanks @doomspork :) There sure were a lot of them

**Comment by doomspork** [comment]
No worries @hassox, always glad to help.  Figure it's easy to point things out now than doing the while PR thing later :)  

I'm really stoked to see this project :+1:

File: README.md [hunk 17]
```diff
@@ -2,6 +2,112 @@
...
+
+Ueberauth provides only the initial authentication challenge. The initial OAuth flow, collecting the information from a login form etc. It does not authenticate each request, that's up to your application. You could issue a token or put the result into a session for your applications needs. Libraries like (Guardian)[https://github.com/hassox/guardian] can help you with that aspect of authentication. Ueberauth manages the initial challenge only.
+
+The two phases are `request` and `callback`. These phases are implemented by Strategies.
+
+### Strategies
+
+Strategies are plugs that decorate or intercept requests (or both).
+Strategies implement the two phases and then may allow the request to flow through to your downstream plugs.
+Implementing the request and callback phases is optional, and if not implemented the request will be decorated Ueberauth information
```
Comment (doomspork) [inline README.md]: I think you're missing a "with" here:  "will be decorated _with_ Ueberauth"

File: README.md [hunk 18]
```diff
@@ -2,6 +2,112 @@
 
 > An Elixir Authentication System for Plug-based Web Applications
 
+Ueberauth is an authentication framework that is heavily inspired by [Omniauth](https://github.com/intridea/omniauth)
```
Comment (doomspork) [inline README.md]: I think you want a period after Omniauth.

**Comment by scrogson** [comment]
@hassox well done!

**Pull request merged**